Option to allow the request body to be processed outside the asynchttpserver library.

mrhdias opened the pull request:
Option to allow the request body to be processed outside the asynchttpserver library to break big files into chunks of data.
This change does not break anything.

Example:
```nim
import asyncnet, asynchttpserver, asyncdispatch

proc test_page(): string =
  return """
<!Doctype html>
<html lang="en">
<head>
<meta charset="utf-8"/>
</head>
<body>
<form action="/test" method="post">
  Test: <input type="text" name="test" value="post test"><br/>
  <input type="submit">
</form>
</body>
</html>
"""

const chunkSize = 1024

var server = newAsyncHttpServer(maxBody=102400, handleBody=false)
proc cb(req: Request) {.async.} =
  if req.reqMethod == HttpPost:
    var body = ""
    var remainder = req.content_length
    while remainder > 0:
      let data = await req.client.recv(if remainder < chunkSize: remainder else: chunkSize)
      if data.len == 0:
        break
      body.add(data)
      remainder -= data.len

    if body.len != req.contentLength:
      await req.respond(Http400, "Bad Request. Content-Length does not match actual.")
    else:
      await req.respond(Http200, body)
    # req.client.close()
  else:
    await req.respond(Http200, test_page())

waitFor server.serve(Port(8080), cb)
```

mrhdias committed:
llow the request body to be processed outside the asynchttpserver library.

Allow the request body to be processed outside the asynchttpserver library to break big files into chunks of data.
This change does not break anything.

Example:

import asyncnet, asynchttpserver, asyncdispatch

proc test_page(): string =
  return """
<!Doctype html>
<html lang="en">
<head>
<meta charset="utf-8"/>
</head>
<body>
<form action="/test" method="post">
  Test: <input type="text" name="test" value="post test"><br/>
  <input type="submit">
</form>
</body>
</html>
"""

const chunkSize = 1024

var server = newAsyncHttpServer(maxBody=102400, handleBody=false)
proc cb(req: Request) {.async.} =
  if req.reqMethod == HttpPost:
    var body = ""
    var remainder = req.content_length
    while remainder > 0:
      let data = await req.client.recv(if remainder < chunkSize: remainder else: chunkSize)
      if data.len == 0:
        break
      body.add(data)
      remainder -= data.len

    if body.len != req.contentLength:
      await req.respond(Http400, "Bad Request. Content-Length does not match actual.")
    else:
      await req.respond(Http200, body)
    # req.client.close()
  else:
    await req.respond(Http200, test_page())

waitFor server.serve(Port(8080), cb)

mrhdias commented:
The example with a minor correction.
```nim
import asyncnet, asynchttpserver, asyncdispatch

proc test_page(): string =
  return """
<!Doctype html>
<html lang="en">
<head>
<meta charset="utf-8"/>
</head>
<body>
<form action="/test" method="post">
  Test: <input type="text" name="test" value="post test"><br/>
  <input type="submit">
</form>
</body>
</html>
"""

const chunkSize = 1024

var server = newAsyncHttpServer(maxBody=102400, handleBody=false)
proc cb(req: Request) {.async.} =
  if req.reqMethod == HttpPost:
    var body = ""
    var remainder = req.content_length
    while remainder > 0:
      let read_size = if remainder < chunkSize: remainder else: chunkSize
      let data = await req.client.recv(read_size)
      if data.len != read_size:
        await req.respond(Http500, "Internal Server Error. An error occurred while reading the request body.")
        return
      body.add(data)
      remainder -= read_size

    if body.len != req.contentLength:
      await req.respond(Http400, "Bad Request. Content-Length does not match actual.")
      return

    await req.respond(Http200, body)
    # req.client.close()
  else:
    await req.respond(Http200, test_page())
    
waitFor server.serve(Port(8080), cb)
```

Araq commented on lib/pure/asynchttpserver.nim:
```diff
@@ -243,10 +247,14 @@ proc processRequest(
-        return true
+
+      request.contentLength = contentLength
+      if server.handleBody == true: # if false leave the body for the developer to do whatever he wants with it.
```
``server.handleBody`` is already of type ``bool`` so the condition should be ``if server.handleBody``.

mrhdias replied:
Ok, you're right

mrhdias committed:
Minor fix suggested by Araq

The line "if server.handleBody == true:" has been replaced by "if server.handleBody:" because "server.handleBody" is already of type bool.
The line "request.contentLength = contentLength" is placed before "if contentLength > server.maxBody:", so it is possible to know the length of the body even if it exceeds the maximum.

mrhdias commented:
Final example without req.contentLength. We can get the contentLength from the header.
```nim
import asyncnet, asynchttpserver, asyncdispatch
import strutils

proc test_page(): string =
  return """
<!Doctype html>
<html lang="en">
<head>
<meta charset="utf-8"/>
</head>
<body>
<form action="/test" method="post">
  Test: <input type="text" name="test" value="post test"><br/>
  <input type="submit">
</form>
</body>
</html>
"""

const chunkSize = 1024

var server = newAsyncHttpServer(maxBody=102400, handleBody=false)
proc cb(req: Request) {.async.} =
  if req.reqMethod == HttpPost:
    var body = ""

    if not req.headers.hasKey("Content-length"):
      await req.respond(Http411, "Content-Length required.") 
      return

    let content_length = req.headers["Content-length"].parseInt

    var remainder = content_length
    while remainder > 0:
      let read_size = if remainder < chunkSize: remainder else: chunkSize
      let data = await req.client.recv(read_size)
      if data.len != read_size:
        await req.respond(Http500, "Internal Server Error. An error occurred while reading the request body.")
        return
      body.add(data)
      remainder -= data.len

    if body.len != contentLength:
      await req.respond(Http400, "Bad Request. Content-Length does not match actual.")
      return

    await req.respond(Http200, body)
    req.client.close()
  else:
    await req.respond(Http200, test_page())
    
waitFor server.serve(Port(8080), cb)
```

mrhdias committed:
No "req.contentLength" required.

No "req.contentLength" required. We can get the contentLength from the header.

dom96 reviewed (changes requested):
I'm not sure we should include something like this in the stdlib, it's kind of a workaround for asynchttpserver not supporting `FutureStream`, maybe that could be implemented instead? (Check out how httpclient does it).

mrhdias committed:
Now, request body is a FutureStream

Example:

import asynchttpserver, asyncdispatch

proc cb(req: Request) {.async.} =
  let htmlpage = """
<!Doctype html>
<html lang="en">
<head>
<meta charset="utf-8"/>
</head>
<body>
<form action="/" method="post" enctype="multipart/form-data">
  File: <input type="file" name="testfile" accept="text/*"><br /><br />
  <input type="submit">
</form>
</body>
</html>
"""

  if req.reqMethod == HttpPost:
    # let body = await req.bodyStream.readAll();
    # echo body

    while (let data = await req.bodyStream.read(); data)[0]:
      echo data[1]

  await req.respond(Http200, htmlpage)

let server = newAsyncHttpServer(maxBody = 1000000000)
waitFor server.serve(Port(8080), cb)

mrhdias commented:
> I'm not sure we should include something like this in the stdlib, it's kind of a workaround for asynchttpserver not supporting `FutureStream`, maybe that could be implemented instead? (Check out how httpclient does it).

I changed the library, now the request body is a FutureStream. Please check this example:
```nim
import asynchttpserver, asyncdispatch

const read_all_body = false

proc cb(req: Request) {.async.} =
  let htmlpage = """
<!Doctype html>
<html lang="en">
<head>
<meta charset="utf-8"/>
</head>
<body>
<form action="/" method="post" enctype="multipart/form-data">
  File: <input type="file" name="testfile" accept="text/*"><br /><br />
  <input type="submit">
</form>
</body>
</html>
"""

  if req.reqMethod == HttpPost:
    if read_all_body:
      # read the request body at once
      let body = await req.bodyStream.readAll();
      echo body
    else:
      # read 1024 bytes at a time
      while (let data = await req.bodyStream.read(); data[0]):
        echo data[1]

  await req.respond(Http200, htmlpage)

let server = newAsyncHttpServer(maxBody = 1000000000)
waitFor server.serve(Port(8080), cb)
```

Araq commented on lib/pure/asynchttpserver.nim:
```diff
@@ -236,17 +245,36 @@ proc processRequest(
+
+      var remainder = contentLength
+      while remainder > 0:
+        let read_size = if remainder < chunkSize: remainder else: chunkSize
```
```suggestion

        let readSize = min(remainder, chunkSize)

```

Araq commented on lib/pure/asynchttpserver.nim:
```diff
@@ -236,17 +245,36 @@ proc processRequest(
+        let read_size = if remainder < chunkSize: remainder else: chunkSize
+        let data = await client.recv(read_size)
+        if data.len != read_size:
+          await request.respond(Http500, "Internal Server Error. An error occurred while reading the request body.")
```
How do you know it's not

```
await request.respond(Http400, "Bad Request. Content-Length does not match actual.")
```

Araq commented on lib/pure/asynchttpserver.nim:
```diff
@@ -236,17 +245,36 @@ proc processRequest(
+        await request.bodyStream.write(data)
+        remainder -= data.len
+
+      if remainder > 0:
```
By construction we know that ``remainder <= 0`` after the while loop.

Araq commented on lib/pure/asynchttpserver.nim:
```diff
@@ -173,17 +190,9 @@ proc processRequest(
   for linePart in lineFut.mget.split(' '):
     case i
     of 0:
-      case linePart
```
I liked the inline ``case`` statement better.

mrhdias replied:
I didn't touch this part of the library. Maybe Dominik Picheta can see this part.

Araq replied:
I think you based your patch on something else but ``nim devel``?

mrhdias replied:
Yes, I based my patch on nim devel.

Araq reviewed (commented):
See my remarks. Also needs a changelog entry and the new public field should be behind a ``.since`` annotation.

mrhdias committed:
Update lib/pure/asynchttpserver.nim

Co-Authored-By: Andreas Rumpf <rumpf_a@web.de>

mrhdias committed:
Added the modifications suggested by Araq.

mrhdias commented:
> See my remarks. Also needs a changelog entry and the new public field should be behind a `.since` annotation.

Can you give me some direction for adding the annotation ".since". I can't find any documentation about it.

mrhdias committed:
The ".since" annotation has been added.

The reading of the body was placed in a separate "proc" for compatibility and because of the ".since" annotation.

mrhdias committed:
The Error: invalid pragma: since: (1, 1) has been fixed.

mrhdias committed:
The ".since" pragma has been removed from the code.

Araq explained on the forum that it is for internal use!

Araq commented on lib/pure/asynchttpserver.nim:
```diff
@@ -52,6 +55,7 @@ type
     url*: Uri
     hostname*: string    ## The hostname of the client that made the request.
     body*: string
+    bodyStream*: FutureStream[string]
```
```nim

when (NimMajor, NimMinor) >= (1, 1):
  bodyStream*: FutureStream[string]
```

mrhdias replied:
> ```nim
> when (NimMajor, NimMinor) >= (1, 1):
>   bodyStream*: FutureStream[string]
> ```
When I put the annotation I get this error:
Error: constant expression expected
```nim
type
  Request* = object
    client*: AsyncSocket # TODO: Separate this into a Response object?
    reqMethod*: HttpMethod
    headers*: HttpHeaders
    protocol*: tuple[orig: string, major, minor: int]
    url*: Uri
    hostname*: string    ## The hostname of the client that made the request.
    body*: string
    when (NimMajor, NimMinor) >= (1, 1):
      bodyStream*: FutureStream[string]
```

Araq replied:
Bah, what's up with that...

mrhdias replied:
Works if outside of type declaration
```nim
when (NimMajor, NimMinor) >= (1, 1):
  const
    maxLine = 8*1024
    chunkSize = 1048

  type
    Request* = object
      client*: AsyncSocket # TODO: Separate this into a Response object?
      reqMethod*: HttpMethod
      headers*: HttpHeaders
      protocol*: tuple[orig: string, major, minor: int]
      url*: Uri
      hostname*: string    ## The hostname of the client that made the request.
      body*: string
      bodyStream*: FutureStream[string] # <<<<<
else:
  const
    maxLine = 8*1024

  type
    Request* = object
      client*: AsyncSocket # TODO: Separate this into a Response object?
      reqMethod*: HttpMethod
      headers*: HttpHeaders
      protocol*: tuple[orig: string, major, minor: int]
      url*: Uri
      hostname*: string    ## The hostname of the client that made the request.
      body*: string

type
  AsyncHttpServer* = ref object
    socket: AsyncSocket
    reuseAddr: bool
    reusePort: bool
    maxBody: int ## The maximum content-length that will be read for the body.
```

mrhdias committed:
Checking versions for compatibility.

Now, check the Nim version to insert new code for compatibility with old versions

mrhdias committed:
Added a entry in library changes for asynchttpserver.

mrhdias committed:
The default value of chunkSize was increased to 8*1024.
